### PR TITLE
User interface

### PR DESCRIPTION
**Pull request by maya88en**
# Milestone 2: User interface

- Suggested branch name: user_interface.

- Display prompts for users, collect user input and only mock the logic (implementing the logic will be the next step).

- Add executable file named main.rb to bin directory. This file should be responsible for user interaction. This is the only one place where puts and get.chomp can be used.

### CODE REVIEW

File: bin/main.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+puts "Enter name for player 2"
+player2 = gets.strip.capitalize
+player2_id = "O"
+
+board = Board.new
+
+Game.new(player1, player2, board).
+
+puts "Would like to play again? [y/ n]"
```
Comment (samgaco) [inline bin/main.rb]: Some things can be added in the mocked logic, refer to the review to see the explanation.

File: bin/main.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+require_relative '../lib/game'
+require_relative '../lib/player'
+require_relative '../lib/board'
+
+puts "Welcome to Tic Tac Toe Game"
+
+puts "This game is played by 2 players, player1 uses 'X' and player2 uses 'O'"
+
+puts "Enter name for player 1"
+player1 = gets.strip.capitalize
```
Comment (samgaco) [inline bin/main.rb]: Any reason to use strip over chomps? No objections against it, just curious :smiley:

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @maya88en and @addod19 :high_brightness: 

- [ ] You should also have into account that **the game logic should be mocked further.** You could, for example, display a board, ask the users to choose a marker ("X" or "O"), show a message indicating who won (randomly), change turns, etc.

**[SUGGESTION]** This depends on how you are thinking about approaching the project,  you have a file named **game_logic.rb** which might not be necessary since you already have **game.rb** 

Work on this and submit back for another code review :+1:  

---------------------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

**Comment by addod19** [comment]
Thank you....We will work on it

**Comment by addod19** [comment]
Hi, how do we Replace class var @@Wins with a class instance var?
It is giving us lint errors.
Thank you.

File: lib/board.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+class Board
+  attr_accessor :grid
+
+  def initialize
+    @grid = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+  end
+
+  def display
+    puts '                       Welcome to Tic Tac Toe Game                  '
```
Comment (bolah2009) [inline lib/board.rb]: #### `REQUIRED`

Let's move all user interaction (`puts`, `print`, `gets.chomp`, `p`, etc) to `main.rb`.

> Add **executable** file named `main.rb` to `bin` directory. **This file should be responsible for user interaction. This is the only one place where `puts` and `get.chomp` (`print` and `p`) can be used.**

File: lib/game.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+require 'set'
+
+class Game
+  @@Wins = [[1, 2, 3], [1, 4, 7], [1, 5, 7], [2, 5, 8], [3, 6, 9], [3, 5, 7], [4, 5, 6], [7, 8, 9]]
```
Comment (bolah2009) [inline lib/game.rb]: Replace `@@Wins` with `@wins` and put in the `initialize` method.

**Review by bolah2009 — CHANGES_REQUESTED** [review]
# Hi Team :smile: 
#### Awesome job so far with the project. :clap: 
A few things to change before moving to the next milestone:
- Fix linter errors
- Use `gets`, `puts`, `print` and `p` in `main.rb` file **only** :relaxed: 
- As this is a collaborative project, please make sure commits shows that. Currently, I can only see one commit from @maya88en.
- Please let's merge PR #1

Kindly, check the comment below for suggestions.

#### :recycle:  Make these changes and submit another code review request. :+1: 
#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

**Comment by addod19** [comment]
any help on how i can fix this kind of error? =>C: Metrics/CyclomaticComplexity: Cyclomatic complexity for game_start is too high. [8/6]
C: Metrics/PerceivedComplexity: Perceived complexity for game_start is too high. [8/7]

File: bin/main.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
+require_relative '../lib/game'
+require_relative '../lib/player'
+
+
+# frozen_string_literal: true
```
Comment (maelfosso) [inline bin/main.rb]: Puts it at the first line

File: bin/main.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+marker2 = gets.chomp.upcase
+
+if marker1 == "X" && marker2 == "O"
+  puts "#{player1} goes with marker #{marker1}"
+  puts "#{player2} goes with marker #{marker2}"
+elsif marker1 == "O" && marker2 == "X"
+  puts "#{player1} goes with marker #{marker1}"
+  puts "#{player2} goes with marker #{marker2}"
+else
+  puts "Enter the right marker!"
```
Comment (maelfosso) [inline bin/main.rb]: After that line, shouldn't you re-ask the marker ?

File: lib/game.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    a = 0
+
+    until board_full?
+      @board.display
+      option = 0
+
+      current_player = a.even? ? @player1 : @player2
+
+      loop do
+        puts " #{current_player}, select an option from 1..9"
```
Comment (maelfosso) [inline lib/game.rb]: Not here

File: lib/game.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        moves[moves.find_index(option)] = char if moves.include?(option)
+        next unless Set.new(moves).empty?
+
+        puts "Game Over, #{current_player} wins!"
+        board.display
+      end
+      @options.push(option)
+      a += 1
+    end
+    puts 'Board FUll, the Game is a tie (draw)'
```
Comment (maelfosso) [inline lib/game.rb]: Not here

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello coders,

Your code is good :smile: 
You did a very very good improvements

But there are things to change.
- [ ] Why do you ask the maker of the two user ? Why not ask one and deduce the other ? Why not fix  it yourself ? I noticed It's possible for the two player to choose the same marker and at the end you're still affecting the mark of your choice.

- [ ] **`REQUIRED`** In that **MILESTONE**, the `lib/` folder should be **EMPTY** !!! **No class** !!!!  Please, **`Rewrite it, with just bin/main.rb`** 

Kindly make those changes and re-submit a new code review request.
Thanks

File: lib/game.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
```
Comment (Ispirett) [inline lib/game.rb]: This should be in the class somewhere in a function that will take the user input and respond based on the user input.

**Review by Ispirett — CHANGES_REQUESTED** [review]
Hey, folks great work with the changes made. Nice readme 👏
 Sadly there are still some bugs :bug:  to squash. 

**Every time a player plays this happens** 

X|X|3
----- ----- -----
O|O|6
----- ----- -----
7|8|9
 Name, select an option from 1..9
3
Welcome to Tic Tac Toe Game
This game is played by 2 players, each player will enter his/her name and choose a marker [X or O]

X|X|X
----- ----- -----
O|O|6
----- ----- -----
7|8|9
 Tim, select an option from 1..9

**When player X,O  won , the game just continued** 

X|X|X
----- ----- -----
O|O|6
----- ----- -----
7|8|9
 Tim, select an option from 1..9

X|X|3
----- ----- -----
O|O|O
----- ----- -----
X|8|9

*I see here you are asking the player if they want to play again but it exits the terminal before we can an option can be chossen*

Board FUll, the Game is a tie (draw)
Would like to play again? [y/ n]

**Also  please fix linter issues**

File: lib/game.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+require 'set'
+
+class Game
+  @wins = [[1, 2, 3], [1, 4, 7], [1, 5, 7], [2, 5, 8], [3, 6, 9], [3, 5, 7], [4, 5, 6], [7, 8, 9]]
```
Comment (adriaanbd) [inline lib/game.rb]: There's no need to repeat `@wins` or options here and then again in your `initialize` method. You can use `attr_writer`, `attr_reader` or `attr_accesor` here. You'd just use one of those three options depending if you just want to read the instance variable (getter), write it (setter) or both (getter and setter). So, you'd only do:

```
attr_accessor :options
attr_reader :wins
```

Read more about that [here](https://mixandgo.com/learn/ruby_attr_accessor_attr_reader_attr_writer).

Comment (adriaanbd) [inline lib/game.rb]: There's no need to repeat `@wins` or options here and then again in your `initialize` method. You can use `attr_writer`, `attr_reader` or `attr_accesor` here. You'd just use one of those three options depending if you just want to read the instance variable (getter), write it (setter) or both (getter and setter). So, you'd only do:

```
attr_accessor :options
attr_reader :wins
```

Read more about that [here](https://mixandgo.com/learn/ruby_attr_accessor_attr_reader_attr_writer).

File: lib/game.rb [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  end
+
+  def board_full?
+    return false if @board.grid.any? { |cell| cell.is_a?(Integer) }
+
+    true
+  end
+
+  def valid_move?(user_input)
+    unless user_input.is_a?(Integer) && ((user_input <= 9) && user_input.positive?) && !@options.include?(user_input)
```
Comment (adriaanbd) [inline lib/game.rb]: There are a lot of conditions in here united by `&&`. If you had an array of possible moves, you could just check for membership with include `@valid_moves.include?(user_input)`

File: lib/game.rb [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+      loop do
+        puts " #{current_player}, select an option from 1..9"
+        option = gets.chomp.to_i
+
+        break if valid_move?(option)
+      end
+
+      char = (a.even? ? 'X' : 'O').to_s
+      @board.grid[option - 1] = char if @board.grid.include?(option)
```
Comment (adriaanbd) [inline lib/game.rb]: You're checking for membership again in here `@board.grid.include?(option)` when your `valid_move` method should be doing that. Either here of over there, but not on both. Your `@board.grid` is already an array of possible moves.

**Comment by addod19** [comment]
fix errors and Complexity error in lint errors

**Comment by addod19** [comment]
Thank you @adriaanbd , one of the reviewers advised me to do
 that...Thanks

File: lib/game.rb [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+      @wins = @wins.each do |moves|
+        moves[moves.find_index(option)] = char if moves.include?(option)
```
Comment (adriaanbd) [inline lib/game.rb]: This is clearly not working, you see when the first player has a winning position, i.e. `[1, 5, 9]`, as such:

```
X|O|3
----- ----- -----
4|X|6
----- ----- -----
O|8|9
```

And player 1 makes the move to `9`, your `@wins` instance variable has the following moves stored:
```
(byebug) @wins
[["X", "O", 3], ["X", 4, "O"], ["X", "X", "O"], ["O", "X", 8], [3, 6, 9], [3, "X", "O"], [4, "X", 6], ["O", 8, 9]]
```

I don't see a triple `X` in there and that is why it isn't recognizing the winner.

You can isolate this on its own and test it by setting `@wins` to whatever you want and see if it determines a winner. Once you have that, you know for sure that this works **if it gets a correct `@wins`** array of arrays; however, from the looks of it, it isn't. Nevertheless, if you check your `@board.grid` the moves are correctly stored there:

```
(byebug) @board.grid
["X", "O", 3, 4, "X", 6, "O", 8, "X"]
```

You can also try to debug it to see what is going on by using `byebug`, all you have to do is `require 'byebug'` at the top of your file and then insert the statement `byebug` where you want to set the breakpoint, then you can use the `s` and `n` commands to step into your code and check the state of your variables at all times.

I have something to ask you though, why are you storing the moves in both `@board.grid` and `@wins`? Surely there must be a better way to go about it. Nevertheless I realize the effort spent, so you just might want to try and figure out why it isn't working to get a working solution out.

**Review by adriaanbd — COMMENTED** [review]
Hi @maya88en I've commented below, hope it helps!

File: bin/main.rb [hunk 14]
```diff
@@ -0,0 +1,48 @@
+
+require_relative '../lib/game'
+require_relative '../lib/board'
+require_relative '../lib/player'
```
Comment (adriaanbd) [inline bin/main.rb:5]: I noticed this class is being required but I don't see it initialized anywhere, are you using yet?

**Review by bolah2009 — APPROVED** [review]
# Hi Team

I am approving this project as **Milestone 2: User interface** :+1: 

Please, merge and submit a Code Review Request for **Milestone 3: Game logic**

File: bin/main.rb [hunk 15]
```diff
@@ -0,0 +1,48 @@
+
```
Comment (bolah2009) [inline bin/main.rb:2]: Let's make `main.rb` an executable file.

Adding `#!/usr/bin/env ruby` to this line will make `bin/main.rb` run as `ruby bin/main.rb` would.
```suggestion
#!/usr/bin/env ruby
# frozen_string_literal: true
```

### Please do this before merging.

**Comment by addod19** [comment]
#!/usr/bin/env ruby added to the main.rb to make it executable.....thanks

**Pull request merged**